packages/grafana-ui/src/components/uPlot/plugins/TooltipPlugin.tsx

```diff
@@ -142,15 +141,93 @@ export const TooltipPlugin: React.FC<TooltipPluginProps> = ({
     tooltip = <SeriesTable series={series} timestamp={xVal} />;
   }
 
-  if (!tooltip || !coords) {
-    return null;
-  }
-
   return (
     <Portal>
-      <VizTooltipContainer position={{ x: coords.viewport.x, y: coords.viewport.y }} offset={{ x: 10, y: 10 }}>
-        {tooltip}
-      </VizTooltipContainer>
+      {tooltip && coords && (
+        <VizTooltipContainer position={{ x: coords.x, y: coords.y }} offset={{ x: TOOLTIP_OFFSET, y: TOOLTIP_OFFSET }}>
+          {tooltip}
+        </VizTooltipContainer>
+      )}
     </Portal>
   );
 };
+
+function isCursourOutsideCanvas({ left, top }: uPlot.Cursor, canvas: DOMRect) {
+  if (left === undefined || top === undefined) {
+    return false;
+  }
+  return left < 0 || left > canvas.width || top < 0 || top > canvas.height;
+}
+
+/**
+ * Given uPlot cursor position, figure out position of the tooltip withing the canvas bbox
+ * Tooltip is positioned relatively to a viewport
+ **/
+function positionTooltip(u: uPlot, bbox: DOMRect) {
```

Comment:
@leeoniya this function does some craaaaazy shit, would love your thoughts on performance (quite a lot of interaction with uPlot API here). Also, not sure yet how to approach testing it, so any ideas are welcome:)

Reply:
We want to render ANY addons via portlal to avoid z-index issues- yes, requirement :)